fix: EXPOSED-277 statementsRequiredToActualizeScheme does not check s…

obabichevjb opened the pull request:
Here is an update for `Table::createStatement()` with the check if the required sequence already exists. If so, it will not create a statement for creating the sequence. The check is based on `Sequence::exists()` method.

The only concern I have is the usage of dialect methods inside `Table` class for checking if something exists or not. It makes `Table::ddl` not idempotent. It might be better to consolidate the logic that says what should be created and what shouldn't depending on real DB, inside SchemaUtils, in the other case if the user simply asks `Table` for statements, they can get different results depending on what they already have in the database.

joc-a commented on exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/ddl/SequencesTests.kt:
```diff
@@ -198,6 +199,29 @@ class SequencesTests : DatabaseTestsBase() {
     }
 
+    @Test
+    fun testABCDEF() {
```
Checking if this test name left like this by mistake?

bog-walk reviewed (changes requested):
I agree about `Table.ddl`. I would suggest to leave everything in `Table` the same, so that the resulting strings from `ddl` etc. do not change depending on what exists in the database.

Instead, I'd isolate the change to `SchemaUtils.createStatements()` in the `flatMap` block so the check only comes into play when one of the actualizatio/migration functions is called. Meaning something that checks the `create` strings and drops them if they create a sequence and if the matching `autoIncSeq` already exists.

obabichevjb commented:
Updated PR:

The main change is that the check of sequence existence moved to SchemaUtils. Table returns a list of statements. But filtering of `create` statements I made via checking if the statement has `create sequence` prefix and sequence name inside. It should be simple enough for now, in another case it's getting more complicated. Some options are to work with `DdlAware` instances on `SchemeUtils` level instead of string statements, or add a public method to `Table` that will allow to get the `Sequnce` instance for this table.

One more small change, which is actually questionable. I moved the field `ddl` from the `Table` and `Column` classes into `DdlAware` interface. It's a question to me why do we need the field `ddl` at all if it simply returns `createStatement()`, but if we want to keep it, it might be a good idea to move it to the interface expecting that we can call `ddl` on any DdlAware instance, not only on `Table`/`Column`

bog-walk commented on exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SchemaUtils.kt:
```diff
@@ -102,13 +102,26 @@ object SchemaUtils {
+        val sequenceExists = autoIncSeq?.let { currentDialect.sequenceExists(Sequence(it)) } ?: false
+
+        return table.ddl.filter { statement ->
+            if (sequenceExists && autoIncSeq != null) {
+                !statement.lowercase().startsWith("create sequence") || !statement.contains(autoIncSeq)
```
If `sequenceExists == true`, `autoIncSeq` could never be null, no? Just to avoid !! in the condition?

obabichevjb replied:
Yep, you're right. I simplified condition a bit to avoid just-type-checking-condition)
